mitodl/ocw-studio
# Website collaboration API

~Temporarily blocked by #73, this PR includes the changes from that one.~

#### Pre-Flight checklist


#### What are the relevant tickets?
Partially addresses #60 

#### What's this PR do?
- Adds an API for collaborator CRUD

#### How should this be manually tested?
Create a `Website` and some users with various permissions in a shell:
```python
from websites.models import Website
from users.models import User
from django.contrib.auth.models import Group

for i in range(1,7):
     User.objects.update_or_create(username=f"fakeuser{i}", email=f"fakeuser{i}@mit.edu", is_active=True)
owner = User.objects.get(username="fakeuser3")
website = Website.objects.create(name="9999_fake_course", title="Fake Course", owner=owner)

User.objects.get(username="fakeuser1").groups.add(website.admin_group)
User.objects.get(username="fakeuser2").groups.add(website.editor_group)
User.objects.get(username="fakeuser4").groups.add(Group.objects.get(name="global_admin"))
User.objects.get(username="fakeuser5").groups.add(Group.objects.get(name="global_author"))
```

If you log in as a superuser or admin for that website (fakuser1, fakeuser4, owner), you should be able to view the URL `http://localhost:8043/api/websites/<website_name>/collaborators/`.  Other users should get a 403.  The page should display a list of users who have permission to view/edit the website, and show the user's username, email, name, plus the group and role they are a member of ('owner' if they are the website owner).

From that api page you should be able to add a new collaborator by sending an email and a role, for example:
```json
{
    "role": "admin",
    "email": "fakeuser1.mit.edu"
}
```

If you try adding a user that already has permission for that website, or use an email that doesn't exist for a user, or try to assign a non-existent group or group for another website, you should get a validation error.

Go to the detail URL for a specific collaborator:  `http://localhost:8043/api/websites/<website_name>/collaborators/<user_username>/`

From here, try POSTing a change from an admin_group to an editor_group or vice versa:
```json
{
    "role": "editor"
}
```

As long as the user is not a global admin or website owner, it should work, otherwise you should get a permission denied error.  You should also get a validation error if you enter an invalid role name.

Click the DELETE button on one of these detail pages.  You should get a 403 if you try it on the site owner or a global admin.  Otherwise, the user should be removed from all permission groups associated with the website.  The user should not be deleted.
